[ci] test reporting

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29497 [ci] test reporting**
* #29608 [jit] change `drop_on_export` warning category

This PR makes our test scripts output artifacts that CircleCI can
understand. This has a few benefits:
1. We can actually see failed tests and their output in the job screen
(instead of having to scroll through logs)
2. We can use the CircleCI test metadata API to track failed tests
programmatically.

it looks like this (old ui): https://circleci.com/gh/pytorch/pytorch/3546584?pipelines-ui-opt-out
or this (new ui): https://app.circleci.com/jobs/github/pytorch/pytorch/3546584/tests


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29497

suo committed:
[wip] test reporting

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

kostmo commented:
## CircleCI build failures summary
As of commit 1902f22:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=1902f2295f99a484ef3ff0210aa375f62f6e3f43).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo committed:
Update on "[wip] test reporting"

suo commented:
it looks like this (old ui): https://circleci.com/gh/pytorch/pytorch/3546584?pipelines-ui-opt-out
or this (new ui): https://app.circleci.com/jobs/github/pytorch/pytorch/3546584/tests

ezyang commented:
Yes please. But what about the distributed tests that are run multiple times with the same name (but different environment variables)?

suo commented on test/test_dataloader.py:
```diff
@@ -283,11 +283,11 @@ def test_iterable_dataset_err(self):
 def set_faulthander_if_available(_=None):
     if HAS_FAULTHANDLER:
-        faulthandler.enable()
+        faulthandler.enable(sys.__stderr__)
```
`xml-test-runner` takes control of the default `sys.stderr`, and their proxy version doesn't have `fileno()` defined, which causes this line to fail. This change just has the fault handler always spit out tracebacks to the "real" `stderr` instead of a proxy.

suo committed:
Update on "[ci] test reporting"


This PR makes our test scripts output artifacts that CircleCI can
understand. This has a few benefits:
1. We can actually see failed tests and their output in the job screen
(instead of having to scroll through logs)
2. We can use the CircleCI test metadata API to track failed tests
programmatically.

it looks like this (old ui): https://circleci.com/gh/pytorch/pytorch/3546584?pipelines-ui-opt-out
or this (new ui): https://app.circleci.com/jobs/github/pytorch/pytorch/3546584/tests


Pull Request resolved: https://github.com/pytorch/pytorch/pull/29497
